File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import "slices"

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}

	root := &TreeNode{Val: preorder[0]}
	i := slices.Index(inorder, root.Val)

	leftInOrder := inorder[:i]
	rigthInOrder := inorder[i+1:]

	leftTreePreOrder := preorder[1 : 1+len(leftInOrder)]
	rightTreePreOrder := preorder[1+len(leftInOrder):]

	root.Left = buildTree(leftTreePreOrder, leftInOrder)
	root.Right = buildTree(rightTreePreOrder, rigthInOrder)

	return root
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */


type Node = TreeNode | null;

class Solution {
  /**
   * @param {Node} root
   * @return {number}
   */

  height(node: Node): number {
    if (node === null) {
      return 0;
    }
    const leftHeight = this.height(node.left);
    const rightHeight = this.height(node.right);
    return Math.max(leftHeight, rightHeight) + 1;
  }

  diameterOfBinaryTree(root: Node): number {
    if (root === null) {
      return 0;
    }
    const leftHeight = this.height(root.left);
    const rightHeight = this.height(root.right);
    const diameterThroughRoot = leftHeight + rightHeight;
    const leftDiameter = this.diameterOfBinaryTree(root.left);
    const rightDiameter = this.diameterOfBinaryTree(root.right);
    return Math.max(diameterThroughRoot, leftDiameter, rightDiameter);
  }
}
